doc/developer/design/20240108_source_metrics_2.md

```diff
@@ -0,0 +1,231 @@
+
+Postgres has two separate definitions:
+
+For the number of rows in each table:
+```
+{table1: 110 rows, table2: 15 rows} - {table1: 100 rows, table2: 10 rows} = 15 rows
+```
+
+And for lsn's during replication:
+```
+100100 lsn - 100050 lsn = 50 bytes
+```
+
+#### MySQL frontier differences
+
+Similar to Postgres, mysql has a rows-per-table difference, as well as a replication difference.
+
+However, during replication, MySQL's source timestamp is partially ordered, instead of just a single
+lsn number:
+
+```
+# source id: gtid
+f1: {0: 1000th gtid, 1: 0th gtid}
+f2: {0: 1010th gtid, 1: 10th gtid}
+```
+In this case, `f2 - f1 = 20 transactions`
+
+#### Comparisons
+
+As you can see, source frontier differences depend on the source, and are primarily useful for _calculating rates_ and when
+_comparing values_. For example:
+
+- If a source's frontier has moved from `{0: 10}` to `{0: 70}` in 1 minute, Materialize can be said to have processed 1 offset/minute.
+- If a source's frontier snapshot frontier is `{table1: 100 rows}` and its current frontier is `{table1: 50}`, Materialize can be said
+to have processed about 50% of the snapshot.
+
+Additionally, differences are designed to be _lower-bounded estimates_. For example, `{0: 100} - {0: 0, 1: 50}` is 100, as we
+choose the lower bound where we assume no offsets past 50 for partition 1 exist.
+
+
+## Snapshot Progress
+
+The first set of metrics this design document proposes involve _snapshot progress_, which is a _lower bound_ on the _percentage_ of the source's snapshot Materialize
+has _read_. These metrics are designed to answer #1 in [the problem statement](#the-problem).
+
+We will introduce 2 new columns in `mz_source_statistics_per_worker`:
+
+```
+| `snapshot_total`     | [`uint8`] | The total number of upstream values that are part of the snapshot. |
+| `snapshot_progress   | [`uint8`] | The number of upstream values Materialize has read so far.         |
+```
+
+The unit of _values_ depends on the source type, and will be _rows_ for MySQL and Postgres, and _offsets_ for kafka.
+
+These values can be summed across workers and compared (`snapshot_progress / snapshot_progress - snapshot_total`) to produce
+a _lower-bound_ estimate on the progress we have made reading the source's snapshot.
+
+### Source frontier difference
+
+Note that this percentage is calculated using _source frontier difference_, between the `snapshot_total` and `snapshot_progress`.
+For consistency within schema of the `mz_source_statistics_per_worker` table, each worker in the source pipeline is required to
+pre-aggregate these values from the possibly complex frontier that represents the total and progress of a snapshot.
+This primarily means summing offsets/rows across partitions/tables.
+
+### Source specifics
+
+The `SourceReader` implementation for each source will be required to produce a `snapshot_total`, as well as
+a continuously updated `snapshot_progress` frontier on each worker.
```

Comment:
It feels like thinking about source frontier differences was a helpful stepping stone to arrive to the realization that we can have each source define its own way of describing its snapshot progress but beyond that they don't offer any advantages implementation wise.

In particular, if the `SourceRender` trait is extended with the requirement that each source provides a `snapshot_total` and a `snapshot_progress` estimate then we don't gain anything from making the type of these two values some weird frontier that will only ever be used by generic code to compute a `u64` difference. For this reason I propose that each source is obliged to present a `u64` typed `snapshot_total` and `snapshot_progress` which it can calculate in whatever way it wants.

In the same vain I propose that each source exposes an `outstanding_total` gauge (again `u64` typed) which represents the number of outstanding "things" upstream. For each source type a thing might be something else, for postgres it might be the number of outstanding LSN bytes and for kafka the number of oustanding messages. In any case though this number going up represents trouble and is a signal to scale up the cluster.

Reply:
I like this! will adjust the doc accordingly